src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
+
+    def estimate_total_duration(self):
+        """
+        Estimates the total time for the zstack acquisition.
+        return (Float > 0): the estimated time
+        """
+        acqui_time = 0.0
+        zstep_time = 0
+        nr_FM_streams = 0
+        for s in self._streams:
+            acqui_time += s.estimateAcquisitionTime() * len(self._zlevels[s])
+            if isinstance(s, FluoStream):
+                nr_FM_streams += 1
+        if self._zpos:
+            zstep_time = self._zstep_duration * (len(self._zpos) - 1) * nr_FM_streams
+        return acqui_time + zstep_time
+
+    def _construct_zcube(self, data_arrays):
+        """
+        Create 3D data array from the 2D acquired data arrays 
+        data_arrays (list of DataArray): a list of acquired images at different z levels.
+        return (DataArray): a data array embedding all the data acquired at all z levels.
+        """
+        # pixel size along z axis
+        metadata = data_arrays[0].metadata
+        px, py = metadata[model.MD_PIXEL_SIZE]
+        pz = (self._zpos[-1] - self._zpos[0]) / (len(self._zpos) - 1)
+
+        if pz < 0:
+            data_arrays = numpy.flipud(data_arrays)
+            pz = -pz
+        # upgrade the metadata with the z pixel size
+        metadata[model.MD_PIXEL_SIZE] = (px, py, round(pz, ndigits=7))
+
+        # z coordinate of center
+        cx, cy = metadata[model.MD_POS]
+        cz = (self._zpos[0] + self._zpos[-1]) / 2
+        # upgrade the metadata with the z center position
+        metadata[model.MD_POS] = (cx, cy, round(cz, ndigits=5))
+
+        return model.DataArray(data_arrays, metadata)
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+        """
+        remaining_t = self.estimate_total_duration()
+        acquired_data = []
+        # iterate through streams
+        for stream in self._streams:
+            zstack = []
+            # for each stream, iterate through zlevels (if exist)
+            for i, z in enumerate(self._zlevels[stream]):
+                # move the focuser
+                self._actuator_f = stream.focuser.moveAbs({"z": z})
+                self._actuator_f.result()
+
+                # check if cancellation happened while the actuator future is working
+                if self._future_state == CANCELLED:
+                    raise CancelledError()
+
+                # if FM stream, then subtract one zstep time
+                if isinstance(stream, FluoStream):
+                    if i != len(self._zpos) - 1:
+                        remaining_t -= self._zstep_duration
+                        self._main_future.set_end_time(time.time() + remaining_t)
+
+                try:
+                    # acquire this single stream, and get the data
+                    self._single_acqui_f = acquire([stream], self._settings_obs)
+                    data, exp = self._single_acqui_f.result()
```

Comment:
You need to do something with the exception `exp`. Basically, if it's not None, you should stop now, pack the data which has already been acquired and return the data (so far) + exp.

Reply:
I'm wondering, if an error happened during the waiting for the result, the exception statement will be executed, right? So i return the data and exception in that statement. Is that what you meant? Please check it in the latest commits.